src/codegen/llvm/llvm_ir_builder.cpp

```diff
@@ -174,6 +174,52 @@ void IRBuilder::create_intrinsic(const std::string& name,
     }
 }
 
+void IRBuilder::set_kernel_attributes() {
+    // By convention, the compute kernel does not free memory and does not throw exceptions.
+    current_function->setDoesNotFreeMemory();
+    current_function->setDoesNotThrow();
+
+    // We also want to specify that the pointers that instance struct holds, do not alias. In order
+    // to do that, we add a `noalias` attribute to the argument. As per Clang's specification:
+    //  > The `noalias` attribute indicates that the only memory accesses inside function are loads
+    //  > and stores from objects pointed to by its pointer-typed arguments, with arbitrary
+    //  > offsets.
+    current_function->addParamAttr(0, llvm::Attribute::NoAlias);
+
+    // Finally, specify that the struct pointer does not capture and is read-only.
+    current_function->addParamAttr(0, llvm::Attribute::NoCapture);
+    current_function->addParamAttr(0, llvm::Attribute::ReadOnly);
+}
+
+/****************************************************************************************/
+/*                                LLVM metadata utilities                               */
+/****************************************************************************************/
+
+void IRBuilder::set_loop_metadata(llvm::BranchInst* branch) {
+    llvm::LLVMContext& context = builder.getContext();
+    MetadataVector loop_metadata;
+
+    // Add nullptr for loop's metadata self-reference.
+    loop_metadata.push_back(nullptr);
```

Comment:
for my understanding as I don't know this - how this `nullptr` for loop's metadata self-reference work? if you could point me some reference / link, I will read about.

Reply:
I think my comment is misleading! I add nullptr to "reserve the first place" for the self-reference. And then later in the code few lines below I call `metadata->replaceOperandWith(0, metadata)` that does the trick.

Reply:
I will adjust the comment specifying what I wrote above.

Reply:
ok, thanks!